### PR TITLE
Allow copying models with a subset of constraints

### PR DESCRIPTION
**Pull request by dourouc05**
This brings https://github.com/jump-dev/MathOptInterface.jl/pull/1135 into JuMP. 

Next step for #1035.

### CODE REVIEW

**Comment by dourouc05** [comment]
I plan to submit https://github.com/dourouc05/JuMP.jl/tree/iis-easy once this is merged, but I'm wondering how to test the new function…

**Comment by dourouc05** [comment]
@mlubin @odow Is this good enough to be merged?

File: src/copy.jl
Comment (mlubin) [inline src/copy.jl:103]: This doesn't catch the case where `value` is a container of ConstraintRefs (please add a test).

Comment (dourouc05) [reply src/copy.jl:103]: I don't get how you can generate such a case (even if it's mentioned in the docs: https://github.com/jump-dev/JuMP.jl/blob/1d36b2c3c1bb132dc2e97af3f2b17ef544eecbca/src/JuMP.jl#L932) from user code. I don't find any occurrence of this in the test suite either.

Comment (mlubin) [reply src/copy.jl:103]: `@constraint(model, constr[i = 1:2], x[i] <= y[i])` for example. See https://jump.dev/JuMP.jl/v0.21.5/constraints/#Constraint-containers-1.

Comment (dourouc05) [reply src/copy.jl:103]: Thanks for your pointer! I'm working on a patch for this.

**Comment by mlubin** [comment]
Please fix the test failure: https://travis-ci.com/github/jump-dev/JuMP.jl/jobs/428464198#L501

```
test_copy_filter: Error During Test at /home/travis/build/jump-dev/JuMP.jl/test/model.jl:520
  Got exception outside of a @test
  MethodError: no method matching getindex(::ReferenceMap, ::Array{VariableRef,1})
  Closest candidates are:
    getindex(::ReferenceMap, !Matched::VariableRef) at /home/travis/build/jump-dev/JuMP.jl/src/copy.jl:27
    getindex(::ReferenceMap, !Matched::ConstraintRef) at /home/travis/build/jump-dev/JuMP.jl/src/copy.jl:31
  Stacktrace:
   [1] copy_model_style_mode(::Bool, ::MathOptInterface.Utilities.CachingOptimizerMode, ::Bool) at /home/travis/build/jump-dev/JuMP.jl/test/model.jl:453
   [2] test_copy_filter() at /home/travis/build/jump-dev/JuMP.jl/test/model.jl:512
```

File: docs/src/installation.md
Comment (mlubin) [inline docs/src/installation.md:43]: How did this diff get picked up?

File: src/copy.jl
Comment (mlubin) [inline src/copy.jl:47]: Typo: argument

File: test/model.jl
Comment (mlubin) [inline test/model.jl:425]: Could you explicitly test all three types of possible containers? They're Arrays, DenseAxisArrays, and SparseAxisArrays: https://jump.dev/JuMP.jl/v0.21.5/constraints/#Constraint-containers-1.

**Comment by dourouc05** [comment]
Normally, failures are now solved (I was testing the wrong version locally…). The other three comments are still outstanding.

File: test/model.jl
Comment (dourouc05) [reply test/model.jl]: I just did that in the latest commit. I was expecting that I could just use an argument to `copy_model_style_mode`, but the `container` argument in both `@variable` and `@constraint` is not evaluated (i.e. the variable name is directly used as a container type, and not the contents of the variable).

File: src/copy.jl
Comment (mlubin) [inline src/copy.jl:184]: Explain more explicitly what the blank line represents. I'm guessing it's for solving the model.

File: test/model.jl
Comment (mlubin) [inline test/model.jl:542]: Wasn't `cref[2]` filtered from the model? Why is it in the reference_map? Same for the two tests below.

File: test/model.jl
Comment (mlubin) [inline test/model.jl:524]: The optimize hook and extension data aren't really relevant for these tests. The same functionality is already tested in copy_model_style_mode. Try to keep the tests focused on one thing, if possible (not all JuMP tests do this).

File: test/model.jl
Comment (mlubin) [inline test/model.jl:530]: Test that the number of AffineExpression-in-EqualTo constraints in the model is one. I think this test will fail because of my comment on `copy_model`.

File: src/copy.jl
Comment (mlubin) [inline src/copy.jl:89]: There's a problem here. `MOI.copy_to` expects the filter to be on `MOI.ConstraintIndex` objects, while a JuMP user is likely to write a filter based on `JuMP.ConstraintIndex` (e.g., see the tests written). Some translation needs to be done.

File: src/copy.jl
Comment (dourouc05) [reply src/copy.jl:184]: It's not necessary to have something there (unless a specific solver requires solving before looking for conflicts?). 

However, I added the same comment around `Model()` to indicate that not all solvers are suitable. Should this example check that the conflict computation could be done properly (i.e. checking `MOI.get(model, MOI.ConflictStatus())`)?

**Comment by dourouc05** [comment]
@mlubin Thanks for having a look! The new commits should fix the issues you raised.

File: test/model.jl
Comment (mlubin) [inline test/model.jl:544]: Please remove `optimize_hook` and `DummyExtensionData` from these tests also.

**Review by mlubin — COMMENTED** [review]
Looks like we're close to merging. Please resolve the merge conflicts so CI can run.

**Comment by dourouc05** [comment]
The decrease in coverage is mostly due to `copy_conflict` not being called in the test suite. I also added a link to the new functionality from `solutions.md` in the docs.

**Review by mlubin — APPROVED** [review]
@odow or @blegat, could you take a quick look?

File: src/copy.jl
Comment (blegat) [inline src/copy.jl:92]: Could you add a comment explaining why this `else` clause is needed ? How could `cref` be something else than a `ConstraintIndex` ?

File: src/copy.jl
Comment (blegat) [inline src/copy.jl:49]: index -> reference

File: src/copy.jl
Comment (blegat) [reply src/copy.jl:92]: I get it, it's because you use it below, maybe do `moi_filter_constraints = (con_idx) -> ...` that only works with index, pass that to `copy_to` and for the object dictionary, use `filter_constraints` ?

**Comment by dourouc05** [comment]
@blegat I have modified the doc. 

@mlubin too: I have added a new test based on MockOptimizer for `copy_conflict`, that should improve both coverage and confidence in the implementation. It needs a few changes in MOI, though: https://github.com/jump-dev/MathOptInterface.jl/pull/1202

File: src/copy.jl
Comment (odow) [reply src/copy.jl:92]:
```Julia
    function moi_filter_constraints(cref::MOI.ConstraintIndex)
        jump_cref = constraint_ref_with_index(model, cref)
        return filter_constraints(jump_cref)
    end
    filter = filter_constraints !== nothing ? moi_filter_constraints : nothing
    
    index_map = MOI.copy_to(
        backend(new_model),
        backend(model),
        copy_names = true,
        filter_constraints = filter
    )
```

File: src/copy.jl
Comment (odow) [inline src/copy.jl:123]: We generally try not to use `isa` like this. How about something like
```Julia
_copy_object(filter, value) = true
_copy_object(filter::Function, value::ConstraintRef) = filter(value)
function _copy_object(filter::Function, value::AbstractArray{<:ConstraintRef})
    return all(filter.(value))
end

for (name, value) in object_dictionary(model)
    if _copy_object(filter_constraints, value)
        new_model.obj_dict[name] = getindex.(reference_map, value)
    end
end
```

What does the `elseif` clause do when we only want some of the constraints? That seems like an issue. We probably want to just skip copying the name into `obj_dict` if only some of the container is copied.

Comment (dourouc05) [reply src/copy.jl:123]: That reasoning would lead to problems when copying just a conflict: maybe only a few constraints of the container are in the conflict, and only those should be copied. I'll propose a new iteration on the code soon.

Comment (odow) [reply src/copy.jl:123]: To clarify: we should copy on the constraints in the conflict, but maybe not set the name in `.obj_dict`.

File: src/copy.jl
Comment (dourouc05) [reply src/copy.jl]: @odow, I used your suggestion in 7601072.

File: src/copy.jl
Comment (dourouc05) [reply src/copy.jl:118]: I think I implemented your suggestion: (1) use functions instead of `isa` and (2) only copy into the dictionary the objects that are completely copied.

**Review by odow — COMMENTED** [review]
Looking good. Can you add `copy_conflict` to https://github.com/jump-dev/JuMP.jl/blob/master/docs/src/reference/solutions.md?

It also deserves a sentence in `docs/src/solutions.md` like: 
```
To create a new model containing only the constraints that participate in the conflict, use [`copy_conflict`](@ref). It may be helpful to write this model to file for easier debugging using [`write_to_file`](@ref).
```

**Comment by dourouc05** [comment]
I did that in the last commit. (`copy_model` was already used in the sample code of `docs/src/solutions.md`.)

**Comment by dourouc05** [comment]
The current test failure is likely only due to the fact that the latest tagged MOI doesn't have https://github.com/jump-dev/MathOptInterface.jl/pull/1202, on which this PR relies for the tests. (Aka: it works on my side®.)

**Review by odow — COMMENTED** [review]
Looks good. There are some style things, I'm slowly rolling out JuliaFormatter, so I'm not worried about them now.

When a new MOI gets tagged, I can sort the rest of this out.

**Comment by dourouc05** [comment]
Great, thanks! 

By the way, I was thinking of implementing CPLEX' and Gurobi's feature to easily write the IIS (i.e. a different extension when writing the files). Shall I do it in this PR or in another one? The problem is that this will depend a lot on `copy_model`.

**Comment by odow** [comment]
You mean as a generic feature? Or specifically for Gurobi/CPLEX?

If we have `copy_conflict` and `write_to_file`, I don't see the benefit of a specialized `write_conflict_to_file`. You can just go:
```Julia
new_model, _ = copy_conflict(model)
write_to_file(new_model, "conflict.lp")
```

**Comment by dourouc05** [comment]
These solvers' API allow you to just write the same as a one-liner (which is nicer to the user): 

```
write_to_file(model, "conflict.clp") # CPLEX extension
write_to_file(model, "conflict.ilp") # Gurobi extension
```

These could also be optimised by avoiding a copy of the model (I'll admit that I didn't do benchmarks to see if it's really worthwhile).

**Comment by odow** [comment]
For the solver-specific API's people can just go
```Julia
model = direct_model(Gurobi.Optimizer())
# stuff
compute_conflict!(model)
GRBwrite(backend(model), "conflict.ilp")
```
In general, if you're debugging an infeasible model, you probably don't care about the extra overhead of the `copy_model`.

**Comment by dourouc05** [comment]
My goal is not to provide access to more of the solvers' API, but only to replicate these features at JuMP's level.

**Comment by dourouc05** [comment]
Too bad it didn't make it to the new version :(.

**Comment by odow** [comment]
Once we make a new MOI release, we can make a new JuMP release. I'm aiming to have more smaller, more frequent releases than the ones we have been having.

**Comment by odow** [comment]
@dourouc05 MOI 0.9.20 has been released, so you can rebase this on the latest master, and bump the minimum version of MOI in `Project.toml`!

**Comment by odow** [comment]
There are still a few conflicts with the documentation

**Comment by dourouc05** [comment]
Here it is, fully rebased :)! I squashed all commits into one to ease rebasing (many conflicts were created by changes from this PR that I ended up reverting in this PR).

**Comment by odow** [comment]
Great! I'll merge this when CI passes. Thanks for sticking with this. It ticks off a big item from our TODO list!

**Comment by dourouc05** [comment]
CI passes, with the exception of docs, but it really looks unrelated (doctest for `modify`).

**Comment by odow** [comment]
That's my fault, sorry. Could you rebase onto latest master?

**Comment by odow** [comment]
Don't worry. I realized I can do it.

**Comment by odow** [comment]
Thanks!

**Comment by dourouc05** [comment]
Thanks for merging this! It's the end of a journey that started nearly four years ago :D!

**Comment by odow** [comment]
And all fixing https://github.com/jump-dev/JuMP.jl/issues/1035#issue-232333727 required was a complete re-write of JuMP, <s>MathProgBase</s>MOI, and CPLEX.jl!

**Pull request merged**
